// igniteui-angular/esm5/lib/calendar/month-picker-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { IgxCalendarBase } from './calendar-base';
import { ViewChild, ElementRef, HostBinding } from '@angular/core';
/** @enum {number} */
var CalendarView = {
    DEFAULT: 0,
    YEAR: 1,
    DECADE: 2,
};
export { CalendarView };
CalendarView[CalendarView.DEFAULT] = 'DEFAULT';
CalendarView[CalendarView.YEAR] = 'YEAR';
CalendarView[CalendarView.DECADE] = 'DECADE';
var IgxMonthPickerBase = /** @class */ (function (_super) {
    tslib_1.__extends(IgxMonthPickerBase, _super);
    function IgxMonthPickerBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * The default `tabindex` attribute for the component.
         *
         * @hidden
         */
        _this.tabindex = 0;
        /**
         * @hidden
         */
        _this._activeView = CalendarView.DEFAULT;
        return _this;
    }
    Object.defineProperty(IgxMonthPickerBase.prototype, "activeView", {
        /**
         * Gets the current active view.
         */
        get: /**
         * Gets the current active view.
         * @return {?}
         */
        function () {
            return this._activeView;
        },
        /**
         * Sets the current active view.
         */
        set: /**
         * Sets the current active view.
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._activeView = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxMonthPickerBase.prototype, "isDefaultView", {
        /**
         * @hidden
         */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this._activeView === CalendarView.DEFAULT;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxMonthPickerBase.prototype, "isDecadeView", {
        /**
         * @hidden
         */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this._activeView === CalendarView.DECADE;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxMonthPickerBase.prototype.changeYear = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        this.viewDate = new Date(event.getFullYear(), this.viewDate.getMonth());
        this._activeView = CalendarView.DEFAULT;
        requestAnimationFrame(function () {
            if (_this.yearsBtn) {
                _this.yearsBtn.nativeElement.focus();
            }
        });
    };
    /**
     * @hidden
     */
    /**
     * @hidden
     * @param {?=} args
     * @return {?}
     */
    IgxMonthPickerBase.prototype.activeViewDecade = /**
     * @hidden
     * @param {?=} args
     * @return {?}
     */
    function (args) {
        this._activeView = CalendarView.DECADE;
    };
    /**
     * @hidden
     */
    /**
     * @hidden
     * @param {?} event
     * @param {?=} args
     * @return {?}
     */
    IgxMonthPickerBase.prototype.activeViewDecadeKB = /**
     * @hidden
     * @param {?} event
     * @param {?=} args
     * @return {?}
     */
    function (event, args) {
        if (event.key === " " /* SPACE */ || event.key === "Spacebar" /* SPACE_IE */ || event.key === "Enter" /* ENTER */) {
            event.preventDefault();
            this.activeViewDecade(args);
        }
    };
    /**
     * Returns the locale representation of the year in the year view if enabled,
     * otherwise returns the default `Date.getFullYear()` value.
     *
     * @hidden
     */
    /**
     * Returns the locale representation of the year in the year view if enabled,
     * otherwise returns the default `Date.getFullYear()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxMonthPickerBase.prototype.formattedYear = /**
     * Returns the locale representation of the year in the year view if enabled,
     * otherwise returns the default `Date.getFullYear()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.formatViews.year) {
            return this.formatterYear.format(value);
        }
        return "" + value.getFullYear();
    };
    IgxMonthPickerBase.propDecorators = {
        yearsBtn: [{ type: ViewChild, args: ['yearsBtn', { static: false },] }],
        tabindex: [{ type: HostBinding, args: ['attr.tabindex',] }]
    };
    return IgxMonthPickerBase;
}(IgxCalendarBase));
export { IgxMonthPickerBase };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    IgxMonthPickerBase.prototype.yearsBtn;
    /**
     * The default `tabindex` attribute for the component.
     *
     * @hidden
     * @type {?}
     */
    IgxMonthPickerBase.prototype.tabindex;
    /**
     * @hidden
     * @type {?}
     * @private
     */
    IgxMonthPickerBase.prototype._activeView;
}
